presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1116,34 +1116,76 @@ protected Scope visitJoin(Join node, Optional<Scope> scope)
+
+                    if (!leftField.isPresent() || !rightField.isPresent()) {
+                        throw missingAttributeException(column, columnName);
+                    }
+
+                    try {
+                        metadata.getFunctionRegistry().resolveOperator(OperatorType.EQUAL, ImmutableList.of(
+                                leftField.get().getType(), rightField.get().getType()));
+                    }
+                    catch (OperatorNotFoundException e) {
+                        throw new SemanticException(TYPE_MISMATCH, node, "%s", e.getMessage());
+                    }
+
+                    Optional<Type> type = metadata.getTypeManager().getCommonSuperType(leftField.get().getType(), rightField.get().getType());
+                    analysis.addTypes(ImmutableMap.of(NodeRef.of(column), type.get()));
 
-                    ExpressionAnalysis leftExpressionAnalysis = analyzeExpression(leftExpression, left);
-                    ExpressionAnalysis rightExpressionAnalysis = analyzeExpression(rightExpression, right);
-                    checkState(leftExpressionAnalysis.getSubqueryInPredicates().isEmpty(), "INVARIANT");
-                    checkState(rightExpressionAnalysis.getSubqueryInPredicates().isEmpty(), "INVARIANT");
-                    checkState(leftExpressionAnalysis.getScalarSubqueries().isEmpty(), "INVARIANT");
-                    checkState(rightExpressionAnalysis.getScalarSubqueries().isEmpty(), "INVARIANT");
+                    joinFields.add(Field.newUnqualified(column.getValue(), type.get()));
 
-                    addCoercionForJoinCriteria(node, leftExpression, rightExpression);
-                    expressions.add(new ComparisonExpression(EQUAL, leftExpression, rightExpression));
+                    leftJoinFields.add(leftField.get().getRelationFieldIndex());
+                    rightJoinFields.add(rightField.get().getRelationFieldIndex());
                 }
 
-                analysis.setJoinCriteria(node, ExpressionUtils.and(expressions));
+                ImmutableList.Builder<Field> builder = ImmutableList.builder();
+                builder.addAll(joinFields);
+
+                List<Integer> leftFields = new ArrayList<>();
+                for (int i = 0; i < left.getRelationType().getAllFieldCount(); i++) {
+                    if (!leftJoinFields.contains(i)) {
+                        builder.add(left.getRelationType().getFieldByIndex(i));
+                        leftFields.add(i);
+                    }
+                }
+
+                List<Integer> rightFields = new ArrayList<>();
+                for (int i = 0; i < right.getRelationType().getAllFieldCount(); i++) {
+                    if (!rightJoinFields.contains(i)) {
+                        builder.add(right.getRelationType().getFieldByIndex(i));
+                        rightFields.add(i);
+                    }
+                }
```

Comment:
this `for` code block is the same for `left` and `right` relation, please extract method

Reply:
It produces two outputs, so hard to do without returning a holder object (more complexity) or having an "out" argument (yuck). Let me play with it a bit.